src/librustc_typeck/outlives/mod.rs

```diff
@@ -23,7 +25,33 @@ pub fn provide(providers: &mut Providers) {
 }
 
 //todo
-fn inferred_outlives_of<'a, 'tcx>(_tcx: TyCtxt<'a, 'tcx, 'tcx>, _def_id: DefId)
+fn inferred_outlives_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
                                   -> Vec<ty::Predicate<'tcx>> {
-    Vec::new()
+
+    if let ty::TyAdt(_def, _) = tcx.type_of(def_id).sty {
```

Comment:
Invoking `type_of` indiscriminanetly is not really ok. What we want to do is to first convert the `def_id` into a local node-id, which will then allow us to query the HIR map to find out what kind of node we are looking at. 

Something like this:

```rust
use hir::map as hir_map;

// Assert that this is a local node-id:
let node_id = self.tcx.as_local_node_id(def_id).unwrap();

match tcx.hir.get(node_id) {
    hir_map::NodeItem(item) => match item.node {
        hir::ItemTrait(..) => ..,
        hir::ItemStruct(..) => ..,
        hir::ItemEnum(..) => ..,
        _ => ...,
    }
    _ => ...
}
```

Here, the `get()` method returns a [`hir::map::Node`](https://github.com/rust-lang/rust/blob/7fa64bcef3a23e2a19792ebe297ed17fab36c273/src/librustc/hir/map/mod.rs#L44-L68). You can see that it has many variants. I think we are just interested in struct/enum here, so that would be `NodeItem`, and then we must further match on the `item.node` to find out what kind of item we have. (As shown above.) For everything that is not a struct or enum, I think we can just return an empty vector.